File: backend/app/models/robot.py
from typing import Literal

from pydantic import BaseModel, Field

MachineState = Literal["idle", "running", "paused", "alarm"]
AlarmSeverity = Literal["info", "warning", "critical"]
SensorValueType = float | int | str


class GantryState(BaseModel):
    x: float
    y: float
    z: float


class GantryStateUpdate(BaseModel):
    x: float | None = None
    y: float | None = None
    z: float | None = None


class SensorState(BaseModel):
    name: str
    value: SensorValueType
    unit: str | None = None


class AlarmState(BaseModel):
    code: str
    message: str
    severity: AlarmSeverity = "warning"
    active: bool = True


class RobotState(BaseModel):
    machine_state: MachineState
    current_workflow: str | None = None
    current_step: int | None = Field(default=None, ge=0)
    gantry: GantryState
    sensor_values: list[SensorState]
    alarms: list[AlarmState]


class RobotStateUpdate(BaseModel):
    machine_state: MachineState | None = None
    current_workflow: str | None = None
    current_step: int | None = Field(default=None, ge=0)
    gantry: GantryStateUpdate | None = None
    sensor_values: list[SensorState] | None = None
    alarms: list[AlarmState] | None = None

File: frontend/src/types/robot.ts
export type MachineState = "idle" | "running" | "paused" | "alarm";
export type AlarmSeverity = "info" | "warning" | "critical";

export interface HealthResponse {
  status: string;
  service: string;
  version: string;
}

export interface CameraStatus {
  available: boolean;
  configured_device: string;
  active_device: string | null;
  width: number | null;
  height: number | null;
  fps: number | null;
  error: string | null;
  stream_url: string | null;
}

export interface GantryState {
  x: number;
  y: number;
  z: number;
}

export interface GantryStateUpdate {
  x?: number;
  y?: number;
  z?: number;
}

export interface SensorState {
  name: string;
  value: number | string;
  unit?: string;
}

export interface AlarmState {
  code: string;
  message: string;
  severity: AlarmSeverity;
  active: boolean;
}

export interface RobotState {
  machine_state: MachineState;
  current_workflow: string | null;
  current_step: number | null;
  gantry: GantryState;
  sensor_values: SensorState[];
  alarms: AlarmState[];
}

export interface RobotStateUpdate {
  machine_state?: MachineState;
  current_workflow?: string | null;
  current_step?: number | null;
  gantry?: GantryStateUpdate;
  sensor_values?: SensorState[];
  alarms?: AlarmState[];
}
